### PR TITLE
Cleanup file access control

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java:387]: why this is better? I would undo this change. It is more clear to use domain objects like schema name here instead of `toString()` representation (which is not intuitive that it is designed for UX message)

Comment (dain) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java:387]: The default method in the super class do this, so this is consistent with other system level items.  I actually would like the connector access control to give absolute names, but they don't really have that information easily.

BTW, I would guess these are written this way because this code was copied from the connector level file access control.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java:174]: you need to also check if user has any access to the schema

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java:489]: you need to also check if user has any full access to the schema or catalog

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/CatalogSchemaAccessControlRule.java
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/CatalogSchemaAccessControlRule.java:31]: I would remove it from here and possibly rename it to `CatalogAccessRulePattern`. Then basing on `CatalogAccessRulePattern` and `SchemaAccessRulePattern` I would build `CatalogSchemaAccessControlRule`. No setters and partially build objects.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java
Comment (dain) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java:174]: `checkAnyTablePermission` checks catalog access internally.  FWIU there is no schema `access` rule, and only a schema `owner` rule.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java
Comment (dain) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java:489]: `checkAnyTablePermission` checks catalog access internally.  FWIU there is no schema `access` rule, and only a schema `owner` rule.

**Review by kokosing — APPROVED** [review]

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java
Comment (dain) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java:174]: I understand now.  I changed this to be `isSchemaOwner() || checkAnyTablePermission()`, so schema owners can see all the table names.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java
Comment (dain) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java:492]: I understand now.  I changed this to be `isSchemaOwner() || checkAnyTablePermission()`, so schema owners can see all the table names.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/CatalogSchemaAccessControlRule.java
Comment (dain) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/CatalogSchemaAccessControlRule.java]: I changed the constructor to just explicitly have arguments of the nested type.  Hopefully we can simplify this later, but it works for now.

File: presto-docs/src/main/sphinx/admin/session-property-managers.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/admin/session-property-managers.rst:72]: Is this a breaking change or do you have code to pull in the old name as well?

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/security/file-system-access-control.rst:42]: catalogs. They do not ..

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/security/file-system-access-control.rst:43]: maybe new paragraph

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/security/file-system-access-control.rst:44]: do we want to omit views ?

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/security/file-system-access-control.rst:46]: matching rule, read from

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/security/file-system-access-control.rst:82]: maybe have an actual section header Visibility ?

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/security/file-system-access-control.rst:83]: two spaces before The..

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [inline presto-docs/src/main/sphinx/security/file-system-access-control.rst:247]: rule, read

**Review by mosabua — CHANGES_REQUESTED** [review]
A bunch of minor suggestions for docs submitted. I have not looked at the code...

File: presto-docs/src/main/sphinx/admin/session-property-managers.rst
Comment (dain) [reply presto-docs/src/main/sphinx/admin/session-property-managers.rst:72]: The old name is still supported via `@JsonAlias`.

Comment (mosabua) [reply presto-docs/src/main/sphinx/admin/session-property-managers.rst:72]: Great .. not sure if we need to even add this as a note to the release notes then.

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (dain) [reply presto-docs/src/main/sphinx/security/file-system-access-control.rst:43]: I like this as an introductory summary paragraph that describe what the rules do at a high level so having these here make sense to me.

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (dain) [reply presto-docs/src/main/sphinx/security/file-system-access-control.rst:82]: I was thinking the section headers would be reserved for the rules sections, but a header seems ok to me

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (mosabua) [reply presto-docs/src/main/sphinx/security/file-system-access-control.rst:43]: Yes .. just add an empty line to break up this large paragraph into two.. but still leave it here..

File: presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedSystemAccessControl.java
Comment (kokosing) [inline presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedSystemAccessControl.java:656]: The thing that we load external file and run tests on it make tests very difficult to follow. Especially when file is reused between tests.

Best would be to have some kind of ddl to create file based system access control with few rules and then test them.

**Review by kokosing — COMMENTED** [review]
LGTM for first 3 commits. Up to `Change catalog check to match other checks`. If you could merge them, then it will be easier to continue with that. It is quite hard for me to digest it all at once.

File: presto-docs/src/main/sphinx/security/file-system-access-control.rst
Comment (dain) [reply presto-docs/src/main/sphinx/security/file-system-access-control.rst:133]: Yes this is how most of the rules today treat these today.  An empty array will always result in "no rule matched", so deny.  For no rules provided, the default is typically allow, but there are exceptions.  For every rule set the docs clearly note what happens.

In the later commits, I normalize catalog, schema, and table rules to this.  Instead of using Optionals for these in the code, I replace Optional.empty() with an allow rule, and the other is just an empty list so they get denied.

**Comment by dain** [comment]
@kokosing I prefer to wait until you are done, and then I will merge... the first couple are simple cleanups, and I don't really want to split the PR.

**Review by kokosing — APPROVED** [review]
Just few comments.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/ImpersonationRule.java
Comment (dain) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/ImpersonationRule.java:37]: Good catch.  Fixed.

**Comment by mosabua** [comment]
👍

**Pull request merged**
